### PR TITLE
Override AutoParamDataSource so it can handle more than 8 shadow maps.

### PR DESCRIPTION
**Pull request by mogumbo**
Ogre will render more than 8 shadow maps (I do not know if there is a limit), but AutoParamDataSource will only deliver texture_viewproj_matrix uniforms to shaders for up to 8 shadow maps. I worked around this deficiency by overriding much of AutoParamDataSource and creating a custom SceneManager and associated factory that can be used to inject it into Gazebo.

### CODE REVIEW

**Comment by scpeters** [comment]
the Ubuntu and macOS CI jobs are using ogre 1.9.5 and look fine, while the windows CI is using ogre 1.10.12 and is currently failing as many of the methods that are `virtual` in [the 1.9 version of OgreAutoParamDataSource.h](https://github.com/OGRECave/ogre/blob/v1.9.1/OgreMain/include/OgreAutoParamDataSource.h#L128-L260) have changed and are [not virtual in 1.10.12](https://github.com/OGRECave/ogre/blob/v1.10.12/OgreMain/include/OgreAutoParamDataSource.h#L127-L262). I think some OGRE version macros could be useful here (`#if OGRE_VERSION_MAJOR == 1 && OGRE_VERSION_MINOR < 10`)

* [Build Status](https://build.osrfoundation.org/job/gazebo-ci-pr_any-windows7-amd64/3158/) https://build.osrfoundation.org/job/gazebo-ci-pr_any-windows7-amd64/3158/

**Comment by iche033** [comment]
nice, I extended `spotlight_shadows_demo.world` to have 8 spot lights + 1 directional light in the [iche033/extend_spotlight_shadows_demo](https://github.com/osrf/gazebo/commit/718b6a859df911252621d03c11116478caf0ed7e) branch. It seems like it's able to access the spot light shadow textures.

File: gazebo/rendering/CustomAutoParamDataSource.cc [hunk 1]
```diff
@@ -0,0 +1,284 @@
+/*
+ * Copyright (C) 2017 Open Source Robotics Foundation
```
Comment (iche033) [inline gazebo/rendering/CustomAutoParamDataSource.cc]: 2021

Comment (iche033) [reply gazebo/rendering/CustomAutoParamDataSource.cc]: and also other new files that are added

File: gazebo/rendering/CMakeLists.txt [hunk 2]
```diff
@@ -23,6 +23,7 @@ set (sources
   COMVisual.cc
   ContactVisual.cc
   Conversions.cc
+  CustomAutoParamDataSource.cc
```
Comment (scpeters) [inline gazebo/rendering/CMakeLists.txt]: we can conditionally add the `.cc` file to `sources` and `.hh` file to `headers` based on the `OGRE_VERSION` like how the `deferred_shading` folder is handled [later in this file](https://github.com/osrf/gazebo/blob/gazebo11_11.4.0/gazebo/rendering/CMakeLists.txt#L144-L146)

File: gazebo/rendering/CustomAutoParamDataSource.cc (hunk 1 above)
Comment (mogumbo) [reply gazebo/rendering/CustomAutoParamDataSource.cc]: done

**Comment by iche033** [comment]
>  I extended spotlight_shadows_demo.world (heightmap_shadows.world) to have 8 spot lights + 1 directional light in the iche033/extend_spotlight_shadows_demo branch. It seems like it's able to access the spot light shadow textures.

Added another demo world in my branch to show spot light rendering and shadows over a heightmap, https://github.com/osrf/gazebo/commit/555ece43bb0128168b5e2c4178db6f9b1f80269e, credits go to @mogumbo for actually rendering the spot lights and getting rid of the unwanted shadow artifacts!

heightmap_shadows

File: gazebo/rendering/ScenePrivate.hh [hunk 3]
```diff
@@ -145,6 +148,9 @@ namespace gazebo
       public: std::vector<OculusCameraPtr> oculusCameras;
 #endif
 
+      /// \brief The ogre scene manager factory.
+      public: CustomSceneManagerFactory *managerFactory = nullptr;
```
Comment (scpeters) [inline gazebo/rendering/ScenePrivate.hh:151]: I think this line needs the version macros

File: gazebo/rendering/ScenePrivate.hh [hunk 4]
```diff
@@ -51,6 +51,9 @@ namespace gazebo
 {
   namespace rendering
   {
+#if OGRE_VERSION_MAJOR == 1 && OGRE_VERSION_MINOR <= 9
+    class CustomSceneManagerFactory;
+#endif
```
Comment (scpeters) [inline gazebo/rendering/ScenePrivate.hh]: I think this doesn't technically need version macros

File: gazebo/rendering/ScenePrivate.hh (hunk 3 above)
Comment (mogumbo) [reply gazebo/rendering/ScenePrivate.hh:151]: fixed

File: gazebo/rendering/ScenePrivate.hh (hunk 4 above)
Comment (mogumbo) [reply gazebo/rendering/ScenePrivate.hh]: fixed

File: gazebo/rendering/CMakeLists.txt (hunk 2 above)
Comment (scpeters) [reply gazebo/rendering/CMakeLists.txt]: thanks the windows build is fixed now

[Build Status](https://build.osrfoundation.org/job/gazebo-ci-pr_any-windows7-amd64/3169/) https://build.osrfoundation.org/job/gazebo-ci-pr_any-windows7-amd64/3169/

File: gazebo/rendering/CMakeLists.txt [hunk 5]
```diff
@@ -134,6 +137,9 @@ set (headers
   WrenchVisual.hh
   ogre_gazebo.h
 )
+if (${OGRE_VERSION} VERSION_LESS 1.10)
+  list(APPEND headers "CustomAutoParamDataSource.hh")
```
Comment (scpeters) [inline gazebo/rendering/CMakeLists.txt:141]: does `gazebo/rendering/CustomSceneManager.hh` need to be installed?

Comment (mogumbo) [reply gazebo/rendering/CMakeLists.txt:141]: It works without it, but it doesn't hurt to add it for completeness. Done.

**Comment by scpeters** [comment]
from an external discussion, it sounds like you are building from Ogre from source, so this pull request isn't necessary. I'll close it for now, but we can revisit it in the future if there is interest

**Pull request closed without merging**